torch/csrc/utils/tensor_new.cpp

```diff
@@ -531,9 +531,39 @@ Tensor as_tensor(const Type& type, PyObject* args, PyObject* kwargs) {
   ParsedArgs<3> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.idx == 0) {
+    at::optional<Device> device_opt = r.deviceOptional(2);
+    const Type & type_default = typeWithDefault(r, 1, 2, type);
+    PyObject* data = r.pyobject(0);
     bool type_inference = r.isNone(1);
-    return internal_new_from_data(
-        typeWithDefault(r, 1, 2, type), r.deviceOptional(2), r.pyobject(0), false, false, type_inference);
+    bool is_copy = false;
+
+    if (THPVariable_Check(data)) {
```

Comment:
Why did this code get so much more complicated? The whole reason `internal_from_data` exists is to avoid repeating those things

Reply:
yeah, I put this here mainly was because of the confusion on whether or not `torch.as_tensor(tensor, dtype=dtype, device=device)` should be detached. If yes, than I will still need a way to figure out whether or not a new tensor is copy constructed.